core/file-tree/index.js

```diff
@@ -176,17 +58,179 @@ var writeDataFile = function (callback) {
+    this.threadsNumber = options && options.threads ? options.threads : 1;
+    this.threadQueues = [];
+    this.ptr = 0;
+    for (var i = 0; i < this.threadsNumber; i++) {
+        this.threadQueues.push([]);
+    }
+};
+
+TasksQueue.prototype.push = function(task) {
+    this.threadQueues[this.getQueueNumber()].push(task);
+};
+
+TasksQueue.prototype.getQueueNumber = function() {
+    this.ptr++;
+    if (this.ptr >= this.threadsNumber) {
+        this.ptr = 0;
+    }
+    return this.ptr;
+};
+
+TasksQueue.processTasksSync = function(tasks, callback) {
+    if (!tasks || !tasks.length) {
+        callback && callback();
+        return;
+    }
+    tasks.shift()(function() {
+        TasksQueue.processTasksSync(tasks, callback);
+    });
+};
+
+TasksQueue.prototype.execute = function(callback) {
+    var self = this;
+    var remaining = this.threadQueues.length;
+    this.threadQueues.forEach(function(queue) {
+        TasksQueue.processTasksSync(queue, function() {
+            if (!--remaining) {
+                callback && callback();
             }
+        });
+    });
+};
 
-            if (typeof callback === 'function') callback(err);
+var tasks = new TasksQueue({
+    'threads': 16
+});
+
+/**
+ * @function fillInSpecDataObject - method to fill in spec data.
+ * It create basic fields and initializes async action to fill last
+ * modification date from git log or nodeJS fs.stat data.
+ *
+ * @param {Object} accumulator - empty spec data object to fill
+ * @param {Object} meta - additional information, whitch is needed to create spec data
+ * @param {String} meta.basename - spec file name
+ * @param {String} meta.dirname - spec directory name
+ * @param {String} meta.root - normalized specs root path
+ * @param {String} meta.app - normalized app root path
+ * @param {Object} meta.fileStat - nodeJS fs.stat method result 
+ */
+var fillInSpecDataObject = function(accumulator, meta) {
+    var command = 'git -C ' + global.opts.core.common.pathToUser + ' log -1 --format="%ad" -- ' + dirname;  
+    var specDataObject = accumulator['specFile'] = accumulator['specFile'] || {};
+    var dirname = meta.dirname;
+    var info = {};
+
+    try {
+        info = JSON.parse(fs.readFileSync(dirname + '/' + config.infoFile));
+    } catch(e) {
+        info = {
+            'error': "Cannot parse the file",
+            'path': dirname + '/' + config.infoFile
+        };
+    };
+
+    Object.keys(info).forEach(function(key) {
+        specDataObject[key] = info[key];
+    });
+
+    specDataObject['id'] = dirname.replace(meta.root, '').replace(meta.app, '');
+    specDataObject['url'] = '/' + specDataObject['id'];
+    specDataObject['fileName'] = meta.basename;
+
+    var thumbPath = path.join(dirname, 'thumbnail.png');
+    //TODO: false should be replaced by undefined or empty string
+    specDataObject['thumbnail'] = fs.existsSync(thumbPath) ? thumbPath.replace(meta.root, '') : false;
+    if (config.getFilesDateFromGit) {
```

Comment:
This feature is switched off by default. If it is switched on but git is not installed we`d get timestamp from filestat. I don't think we need to check it in any specific way. Shell I add a kind of warning or error handling for that?